src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp

```diff
@@ -54,6 +107,83 @@ bool ControllerScriptEngineLegacy::callFunctionOnObjects(
             success = false;
         }
     }
+#ifdef MIXXX_USE_QML
+    if (m_bQmlMode) {
+        QHashIterator<QString, std::shared_ptr<QQuickItem>> i(m_rootItems);
+        while (i.hasNext()) {
+            i.next();
```

Comment:
couldn't we simplify by calling methods by "name" just like we do it above in the plain QJSEngine way?

Reply:
No, it's not possible when using QMetaObject::invoke. You must now the exact signature. This is why we need such a complicated fiddling to make it play nice.

Reply:
Right, but shouldn't this be possible without `QMetaObject::invoke` and instead just `qmlEngine->evaluate("rootItem.func(arg1,arg2,etc)")` instead?

Reply:
No because, IIRC, the function isn't part of the global context. here is QML snippet example:

```qml

Item {
    id: root

    function init(controlerName, isDebug) {
    }
}
```

`init` cannot be called in the global context, like it is for Javascript. Instead, it needs to be called on the QML Root item

Reply:
Interesting, good to know. Well in that case it might make sense to simply not support QML methods in `callFunctionOnObjects` and simply invoke methods directly. Or create a small templated variadic function that works better with the variadic implementation of `QMetaMethod::invoke`.

Reply:
Well, that would be quite a regression tho. Currently, these methods allow, like for JS controller, lifecycle synchronisation, such as the critical requirement of cleaning the screen on teardown (the splashoff animation). Without this, the controller will keep the last "live" and look bricked.
> Or create a small templated variadic function that works better with the variadic implementation of QMetaMethod::invoke.

Not supported in Qt 6.2

Reply:
> Well, that would be quite a regression tho. Currently, these methods allow, like for JS controller, lifecycle synchronisation, such as the critical requirement of cleaning the screen on teardown (the splashoff animation).

Yeah yeah I get that. but I'm asking whether we can call these lifecycle methods directly for QML because we know their signature? 

We need `callFunctionOnObjects` for JS because thats how we invoke the handlers whose "address" we don't know. But for QML we know there is only `init` and `shutdown` on the rootItem directly and no handler (we can't even access them as you pointed out). So we could drastically reduce the complexity here by not implementing this dispatch logic. 

> Not supported in Qt 6.2

Ah yikes. Well we could put a better implementation behind a QT_VERSION_CHECK.

Reply:
> Yeah yeah I get that. but I'm asking whether we can call these lifecycle methods directly for QML because we know their signature?

The reason why I decided to make it backward compatible with the JS one is because I thought it could be nice to allow some mapping to be exclusively in QML in the future (similar to Traktor), even if there is no screen rendering. It felt like this would be the natural direction that we would take to allow module controller (like the S4 and its two devices) to run on a single mapping. I guess if it isn't something we would want to allow or encourage in the future, it makes sense to remove it. Also, this file is still called `Legacy` so I guess it makes sense to not think to far ahead. Wdyt?

> we could put a better implementation behind a QT_VERSION_CHECK.

I started doing so but since I'm unable to test it at the moment, it's kinda doing it in the dark. Are you okay with a `TODO` so I or someone else can revisit it when they have a Qt 6.5 available?

Reply:
> The reason why I decided to make it backward compatible with the JS one is because I thought it could be nice to allow some mapping to be exclusively in QML in the future (similar to Traktor), even if there is no screen rendering. It felt like this would be the natural direction that we would take to allow module controller (like the S4 and its two devices) to run on a single mapping. 

That is definitely still on the table (its a potential topic for my bachelors thesis). But the current way of doing input handling is not compatible with QML (only the solution introduced in #12781 is, and that one doesn't use `callFunctionOnObjects`). So I don't see a good reason to support QML here. 

> I guess if it isn't something we would want to allow or encourage in the future, it makes sense to remove it. Also, this file is still called Legacy so I guess it makes sense to not think to far ahead. Wdyt?

I mean it does make sense to think ahead a little bit, but rather in terms of how we can keep new features decoupled from the legacy classes so its easier to port to a future implementation such as `ControllerModuleEngine`. I wouldn't rely on `*Legacy` being around for ever. 

> I started doing so but since I'm unable to test it at the moment, it's kinda doing it in the dark. Are you okay with a TODO so I or someone else can revisit it when they have a Qt 6.5 available?

Sure sure. I can also take a stab at it when I find the time. I'm on Qt 6.7. But at the same time, I really don't think its needed ^^.

Reply:
I'll try and make `ControllerScriptEngineLegacy::extractTransformFunction` templated so it can be used for `shutdown` and `init`

Reply:
Yeah something like that. Don't need to reach for templates though, you can just take `kScreenTransformFunctionUntypedSignature` as a variable and return the `QMetaMethod` instead of inserting it into the hash.